### PR TITLE
[null-safety] implement autodetection for the web

### LINKED ISSUE #69416: Adjust auto-detect logic for flutter web null safety
In https://github.com/flutter/flutter/pull/69405 I will land an approximation of the sound/unsound mode detection to move the roll through.

The "real" fix needs to check:

1. 2.12 language version or 2.11/2.10 language version for opted-in packages (Latter might be ignorable)
2. language version of entry-point library.

This does not need to run on mobile/desktop which can use CFE auto-detection for sound mode

### LINKED ISSUE #70121: Web: change SDK selection when NullSafetyMode.autodetect is working
From @jonahwilliams 

> [NullSafetyMode.autodetect is when] the user did not pass a null safety flag
> so what we should do is look at the SDK version and entrypoint to determine if it is opted in
> but that is not implemented so it falls back to unsound on web

We didn't fall back to unsound. So in https://github.com/flutter/flutter/pull/70120 I changed it to actually fall back to the unsound SDK.

### PR DESCRIPTION
**Pull request by jonahwilliams**
## Description

Fixes https://github.com/flutter/flutter/issues/69416
Fixes https://github.com/flutter/flutter/issues/70121

Update determineLanguageVersion to read the language version, not just as a comment. Use this to determine which runtime mode to support on the web, removing the use of autodetect enum entirely

This logic needs to expanded in the future to support the language version queries and null safety analytics

### CODE REVIEW

**Comment by jonahwilliams** [comment]
Autodetect doesn't work the same way on the web because the tool is ultimately responsible for 1) providing a separate sound or unsound dill 2) serving the precompiled sound dart_sdk.js or unsound dart_sdk.js

File: packages/flutter_tools/lib/src/resident_runner.dart [hunk 1]
```diff
@@ -101,14 +103,28 @@ class FlutterDevice {
...
+          entrypointFile,
+          packageConfig.packageOf(entrypointFile.absolute.uri),
+        );
+        if (languageVersion.major >= nullSafeVersion.major && languageVersion.minor >= nullSafeVersion.minor) {
+          platformDillArtifact = Artifact.webPlatformSoundKernelDill;
+          extraFrontEndOptions =  <String>[
+            ...?buildInfo.extraFrontEndOptions,
+            '--sound-null-safety',
+          ];
+        } else {
```
Comment (zanderso) [inline packages/flutter_tools/lib/src/resident_runner.dart:121]: Check indentation.

**Review by zanderso — APPROVED** [review]

File: packages/flutter_tools/lib/src/resident_runner.dart (hunk 1 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/resident_runner.dart:121]: Fixed

**Review by yjbanov — APPROVED** [review]
lgtm

File: packages/flutter_tools/lib/src/dart/language_version.dart [hunk 2]
```diff
@@ -5,22 +5,24 @@
...
 
-final RegExp _languageVersion = RegExp(r'\/\/\s*@dart');
+final RegExp _languageVersion = RegExp(r'\/\/\s*@dart\s*=\s*([0-9])\.([0-9]+)');
 final RegExp _declarationEnd = RegExp('(import)|(library)|(part)');
 const String _blockCommentStart = '/*';
 const String _blockCommentEnd = '*/';
 
-/// Attempts to read the language version of a dart [file], returning
-/// the entire comment.
+/// The first language version where null safety was available by default
```
[the remark this answers is not in the document]
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/dart/language_version.dart:13]: Language versions aren't sound or unsound. Sound = all libraries have a language version gte 2.12

**Pull request merged**